pdns/dns_random.cc

```diff
@@ -22,73 +22,263 @@
 #ifdef HAVE_CONFIG_H
 #include "config.h"
 #endif
-#include <openssl/aes.h>
-#include <openssl/opensslv.h>
-#if OPENSSL_VERSION_NUMBER > 0x1010000fL && !defined LIBRESSL_VERSION_NUMBER
-// Older OpenSSL does not have CRYPTO_ctr128_encrypt. Before 1.1.0 the header
-// file did not have the necessary extern "C" wrapper. In 1.1.0, AES_ctr128_encrypt
-// was removed.
-#include <openssl/modes.h>
-#endif
-#include <iostream>
-#include <cstdlib>
-#include <cstring>
 #include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
 #include <unistd.h>
-#include <sys/time.h>
-#include <limits>
-#include <stdexcept>
-#include <stdint.h>
+#include <stdlib.h>
+#include <string>
 #include "dns_random.hh"
+#include "arguments.hh"
+#include "logger.hh"
+#include "boost/lexical_cast.hpp"
+
+#if defined(HAVE_RANDOMBYTES_STIR)
+#include <sodium.h>
+#endif
+#if defined(HAVE_RAND_BYTES)
+#include <openssl/rand.h>
+#endif
+#if defined(HAVE_GETRANDOM)
+#include <linux/random.h>
+#endif
 
-using namespace std;
+static enum DNS_RNG {
+  RNG_UNINITIALIZED = 0,
+  RNG_SODIUM,
+  RNG_OPENSSL,
+  RNG_GETRANDOM,
+  RNG_ARC4RANDOM,
+  RNG_URANDOM,
+  RNG_KISS,
+} chosen_rng = RNG_UNINITIALIZED;
 
-static AES_KEY aes_key;
-static unsigned int g_offset;
-static unsigned char g_counter[16], g_stream[16];
-static uint32_t g_in;
+static int urandom_fd = -1;
+
+/* KISS is intended for development use only */
+static unsigned int kiss_seed;
```

Comment:
i'd argue it should be under a configure flag `--dangerous-unsafe-rng`

Reply:
Agreed, or even an `#ifdef 0` so that you have to edit the code to be able to compile it in.